core/state/node/transitions.go

```diff
@@ -109,25 +118,46 @@ func (t *MigrationTransition) Validate(oldState []byte) error {
 }
 
 type AddUserTransition struct {
-	UserID      string `json:"user_id"`
 	Username    string `json:"username"`
 	Certificate string `json:"certificate"`
+
+	EnrichedData *AddUserTranstitionEnrichedData `json:"enriched_data"`
+}
+
+type AddUserTranstitionEnrichedData struct {
+	User *User `json:"user"`
 }
 
 func (t *AddUserTransition) Type() string {
 	return TransitionAddUser
 }
 
 func (t *AddUserTransition) Patch(oldState []byte) ([]byte, error) {
+
+	user, err := json.Marshal(t.EnrichedData.User)
+	if err != nil {
+		return nil, err
+	}
+
 	return []byte(fmt.Sprintf(`[{
 		"op": "add",
 		"path": "/users/%s",
-		"value": {
-			"id": "%s",
-			"username": "%s",
-			"certificate": "%s"
-		}
-	}]`, t.UserID, t.UserID, t.Username, t.Certificate)), nil
+		"value": %s
+	}]`, t.EnrichedData.User.ID, user)), nil
+}
+
+func (t *AddUserTransition) Enrich(oldState []byte) error {
+
+	id := uuid.New().String()
+
+	t.EnrichedData = &AddUserTranstitionEnrichedData{
```

Comment:
I'm a little confused by `Enrich` because it sounds like "metadata" or extra info ... why can't this data be generated at the time the transition is committed and then returned to the caller?  like maybe in an `Apply` method

Reply:
Definitely open to suggestions on naming, I'm not sure if `Enrich` is the best we can do.

The main purpose it to clearly separate out the information needed to request a state update, from the schema for the data that actually makes up a given state update. Having these separated will help prevent a lot of bugs.

I need to think about how to return the data to users. It would be useful in some cases, but it runs into issues with the asynchronous way requests to Raft have to take place. If we make the function blocking so we can return the data after application, we'd could potentially wait indefinitely if Raft gets into a bad state. 

But, I think we can potentially return the data at the request time, which could be helpful in some situations.

Reply:
Hm yeah still not convinced having extra enriched data fields is the right move, but i understand the two different use cases now, i think it's fine for a while